external/kafka-0-10/src/main/scala/org/apache/spark/streaming/kafka/CachedKafkaConsumer.scala

```diff
@@ -0,0 +1,184 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one or more
+ * contributor license agreements.  See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License.  You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.spark.streaming.kafka
+
+import java.{ util => ju }
+
+import org.apache.kafka.clients.consumer.{ ConsumerConfig, ConsumerRecord, KafkaConsumer }
+import org.apache.kafka.common.TopicPartition
+
+import org.apache.spark.SparkConf
+import org.apache.spark.internal.Logging
+
+
+/**
+ * Consumer of single topicpartition, intended for cached reuse.
+ * Underlying consumer is not threadsafe, so neither is this,
+ * but processing the same topicpartition and group id in multiple threads is usually bad anyway.
```

Comment:
concurrentJobs is still undocumented right?  I wouldn't expect the existing direct stream to behave predictably under those conditions either.

The only time I seeing trying to process different chunks of the same topicpartition at the same time as being unavoidable is during checkpoint recovery, which is why the cache isn't used then.

If someone has a better idea I'm all ears, but given the prefetching / handshaking implications of the new consumer, I don't see an alternative to caching them.  We could write considerably more complicated caching logic with a checkout / return system where a certain number of consumers for the same topicpartition were allowed at a time... but IMHO multiple consumers of the same topicpartition in the same job in the same jvm is not something we want to encourage - it's at best a waste of resources and at worst a correctness problem.